tsdb/engine/tsm1/compact.go

```diff
@@ -1282,53 +1284,120 @@ func (k *tsmKeyIterator) Close() error {
 type cacheKeyIterator struct {
 	cache *Cache
 	size  int
+	order []string
 
+	i      int
+	blocks [][]cacheBlock
+	ready  []chan struct{}
+}
+
+type cacheBlock struct {
 	k                string
-	order            []string
-	values           []Value
-	block            []byte
-	minTime, maxTime time.Time
+	minTime, maxTime int64
+	b                []byte
 	err              error
 }
 
 // NewCacheKeyIterator returns a new KeyIterator from a Cache.
 func NewCacheKeyIterator(cache *Cache, size int) KeyIterator {
 	keys := cache.Keys()
 
-	return &cacheKeyIterator{
-		size:  size,
-		cache: cache,
-		order: keys,
+	chans := make([]chan struct{}, len(keys))
+	for i := 0; i < len(keys); i++ {
+		chans[i] = make(chan struct{}, 1)
+	}
+
+	cki := &cacheKeyIterator{
+		i:      -1,
+		size:   size,
+		cache:  cache,
+		order:  keys,
+		ready:  chans,
+		blocks: make([][]cacheBlock, len(keys)),
+	}
+	go cki.encode()
+	return cki
+}
+
+func (c *cacheKeyIterator) encode() {
+	concurrency := runtime.NumCPU()
```

Comment:
Shouldn't this be `runtime.GOMAXPROCS(0)` instead of `runtime.NumCPU`, to respect the `GOMAXPROCS` environment setting?

As a side note, there's a few other places in the tsm1 package where we're using `NumCPU`. I think those ought to all be `GOMAXPROCS`.